### PR TITLE
Fix some panic cuased by nil block, statedb, header

### PR DESCRIPTION
**Pull request by gzliudan**
# Proposed changes

This PR fixes some panic caused by nil block, header, statedb, such as #271.

## Types of changes

What types of changes does your code introduce to XDC network?
_Put an `✅` in the boxes that apply_

- [X] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Regular KTLO or any of the maintaince work. e.g code style
- [ ] CICD Improvement

## Impacted Components
Which part of the codebase this PR will touch base on,

_Put an `✅` in the boxes that apply_

- [ ] Consensus
- [ ] Account
- [ ] Network
- [ ] Geth
- [ ] Smart Contract
- [ ] External components
- [X] Not sure (Please specify below)

## Checklist
_Put an `✅` in the boxes once you have confirmed below actions (or provide reasons on not doing so) that_

- [X] This PR has sufficient test coverage (unit/integration test) OR I have provided reason in the PR description for not having test coverage
- [ ] Provide an end-to-end test plan in the PR description on how to manually test it on the devnet/testnet.
- [ ] Tested the backwards compatibility.
- [ ] Tested with XDC nodes running this version co-exist with those running the previous version.
- [ ] Relevant documentation has been updated as part of this PR
- [ ] N/A

### CODE REVIEW

File: eth/hooks/engine_v1_hooks.go [hunk 1]
```diff
@@ -228,11 +228,11 @@ func AttachConsensusV1Hooks(adaptor *XDPoS.XDPoS, bc *core.BlockChain, chainConf
 		)
 
 		stateDB, err := bc.StateAt(bc.GetBlockByHash(block).Root())
-		candidateAddresses = state.GetCandidates(stateDB)
-
-		if err != nil {
+		if stateDB == nil || err != nil {
 			return nil, err
```
Comment (wgr523) [inline eth/hooks/engine_v1_hooks.go:232]: if stateDB == nil and err == nil, this `return nil, err` will return nil,nil

File: internal/ethapi/api.go [hunk 2]
```diff
@@ -893,7 +893,7 @@ func (s *PublicBlockChainAPI) GetCandidateStatus(ctx context.Context, coinbaseAd
 		candidates, err = s.getCandidatesFromSmartContract()
 	} else {
 		statedb, _, err := s.b.StateAndHeaderByNumber(ctx, checkpointNumber)
-		if err != nil {
+		if statedb == nil || err != nil {
 			result[fieldSuccess] = false
 			return result, err
```
Comment (wgr523) [inline internal/ethapi/api.go:898]: if stateDB == nil and err == nil, this `return` will return result ,nil

File: internal/ethapi/api.go [hunk 3]
```diff
@@ -1048,7 +1048,7 @@ func (s *PublicBlockChainAPI) GetCandidates(ctx context.Context, epoch rpc.Epoch
...
```
Comment (wgr523) [inline internal/ethapi/api.go:1053]: if stateDB == nil and err == nil, this `return` will return result,nil

File: internal/ethapi/api.go [hunk 4]
```diff
@@ -3610,11 +3616,11 @@ func GetSignersFromBlocks(b Backend, blockNumber uint64, blockHash common.Hash,
 		}
 		for i := blockNumber + 1; i <= limitNumber; i++ {
 			header, err := b.HeaderByNumber(nil, rpc.BlockNumber(i))
-			if err != nil {
+			if header == nil || err != nil {
 				return addrs, err
```
Comment (wgr523) [inline internal/ethapi/api.go:3620]: if header == nil and err == nil, this `return` will return addrs,nil

File: internal/ethapi/api.go [hunk 5]
```diff
@@ -3610,11 +3616,11 @@ func GetSignersFromBlocks(b Backend, blockNumber uint64, blockHash common.Hash,
...
 			}
 			blockData, err := b.BlockByNumber(nil, rpc.BlockNumber(i))
-			if err != nil {
+			if blockData == nil || err != nil {
 				return addrs, err
```
Comment (wgr523) [inline internal/ethapi/api.go:3624]: if blockData == nil and err == nil, this `return` will return addrs,nil

**Review by wgr523 — APPROVED** [review]

**Review by wgr523 — APPROVED** [review]

File: eth/api_backend.go [hunk 6]
```diff
@@ -533,11 +533,14 @@ func (b *EthApiBackend) GetEpochDuration() *big.Int {
 func (b *EthApiBackend) GetMasternodesCap(checkpoint uint64) map[common.Address]*big.Int {
 	checkpointBlock := b.eth.blockchain.GetBlockByNumber(checkpoint)
 	state, err := b.eth.blockchain.StateAt(checkpointBlock.Root())
-
 	if err != nil {
 		fmt.Println("ERROR Trying to getting state at", checkpoint, " Error ", err)
 		return nil
 	}
+	if state == nil {
+		fmt.Println("nil state in GetMasternodesCap at checkpoint", checkpoint)
```
Comment (liam-lai) [inline eth/api_backend.go:541]: can you help me use `log.Error` instead of `fmt.Println` please, together with above err condition.

**Review by wjrjerome — APPROVED** [review]

**Review by liam-lai — APPROVED** [review]

**Review by liam-lai — APPROVED** [review]

**Pull request merged**